model/task/task.go

```diff
@@ -871,35 +884,240 @@ func (t *Task) ActivateTask(caller string) error {
+	for _, t := range sortedDependencies {
+		if t.Activated || !t.DeactivatedForDependency {
+			continue
+		}
+
+		depsSatisfied := true
+		for _, dep := range t.DependsOn {
+			// not being activated now
+			if _, ok := tasksToActivate[dep.TaskId]; !ok && !taskMap[dep.TaskId] {
+				// and not already activated
+				if depTask := missingTaskMap[dep.TaskId]; !depTask.Activated {
+					depsSatisfied = false
+					break
+				}
+			}
+		}
+		if depsSatisfied {
+			tasksToActivate[t.Id] = t
+		}
+	}
+
+	if len(tasksToActivate) == 0 {
+		return nil
+	}
+
+	taskIDsToActivate := make([]string, 0, len(tasksToActivate))
+	for _, t := range tasksToActivate {
+		taskIDsToActivate = append(taskIDsToActivate, t.Id)
+	}
+	_, err = UpdateAll(
+		bson.M{IdKey: bson.M{"$in": taskIDsToActivate}},
+		bson.M{"$set": bson.M{
+			ActivatedKey:                true,
+			DeactivatedForDependencyKey: false,
+			ActivatedByKey:              caller,
+			ActivatedTimeKey:            time.Now(),
+		}},
+	)
+
+	for _, t := range tasksToActivate {
+		event.LogTaskActivated(t.Id, t.Execution, caller)
+	}
+
+	return errors.Wrap(err, "can't update activation for dependencies")
+}
+
+func topologicalSort(tasks []Task) ([]Task, error) {
+	depGraph := simple.NewDirectedGraph()
+	taskNodeMap := make(map[string]graph.Node)
+	nodeTaskMap := make(map[int64]Task)
+
+	for _, task := range tasks {
+		node := depGraph.NewNode()
+		depGraph.AddNode(node)
+		nodeTaskMap[node.ID()] = task
+		taskNodeMap[task.Id] = node
+	}
+
+	for _, task := range tasks {
+		for _, dep := range task.DependsOn {
+			if toNode, ok := taskNodeMap[dep.TaskId]; ok {
+				edge := simple.Edge{
+					F: simple.Node(toNode.ID()),
+					T: simple.Node(taskNodeMap[task.Id].ID()),
+				}
+				depGraph.SetEdge(edge)
+			}
+		}
+	}
+
+	sorted, err := topo.Sort(depGraph)
+	if err != nil {
+		return nil, errors.Wrap(err, "problem with topological sort")
+	}
+	sortedTasks := make([]Task, 0, len(tasks))
+	for _, node := range sorted {
+		sortedTasks = append(sortedTasks, nodeTaskMap[node.ID()])
+	}
+
+	return sortedTasks, nil
 }
 
 // DeactivateTask will set the ActivatedBy field to the caller and set the active state to be false and deschedule the task
 func (t *Task) DeactivateTask(caller string) error {
 	t.ActivatedBy = caller
 	t.Activated = false
 	t.ScheduledTime = utility.ZeroTime
-	return UpdateOne(
+
+	return DeactivateTasks([]Task{*t}, caller)
+}
+
+func DeactivateTasks(tasks []Task, caller string) error {
+	taskIDs := make([]string, 0, len(tasks))
+	for _, t := range tasks {
+		taskIDs = append(taskIDs, t.Id)
+	}
+
+	_, err := UpdateAll(
 		bson.M{
-			IdKey: t.Id,
+			IdKey: bson.M{"$in": taskIDs},
 		},
 		bson.M{
 			"$set": bson.M{
 				ActivatedKey:     false,
+				ActivatedByKey:   caller,
 				ScheduledTimeKey: utility.ZeroTime,
 			},
 		},
 	)
+	if err != nil {
+		return errors.Wrap(err, "problem deactivating tasks")
+	}
+
+	for _, t := range tasks {
+		event.LogTaskDeactivated(t.Id, t.Execution, caller)
```

Comment:
In DeactivateDependencies you log regardless of error, we should be consistent about when we do this.

Reply:
I think this is the point I don't follow.
I thought DeactivateDependencies only logs when the update was successful.

Reply:
Ah in DeactivateDependencies the error isn't checked until after the loop that logs tasks being deactivated.